File: backend/app/api/goals.py
from typing import Any, Dict, List, Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from ..database import get_connection
from .auth import get_current_user

router = APIRouter()


class GoalCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = ""
    category: Optional[str] = "general"
    target_date: Optional[str] = None


class GoalUpdateRequest(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = None
    status: Optional[str] = None
    target_date: Optional[str] = None


@router.get("", response_model=List[Dict[str, Any]])
async def list_goals(current_user: Dict[str, Any] = Depends(get_current_user)) -> List[Dict[str, Any]]:
    user_id = current_user["id"]
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM goals WHERE user_id = ? ORDER BY id ASC", (user_id,))
    rows = cursor.fetchall()
    conn.close()
    return [dict(row) for row in rows]


@router.post("", response_model=Dict[str, Any])
async def create_goal(
    goal_in: GoalCreateRequest,
    current_user: Dict[str, Any] = Depends(get_current_user),
) -> Dict[str, Any]:
    user_id = current_user["id"]
    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute(
        """
        INSERT INTO goals (user_id, title, description, category, status, target_date)
        VALUES (?, ?, ?, ?, 'active', ?)
        """,
        (
            user_id,
            goal_in.title,
            goal_in.description,
            goal_in.category or "general",
            goal_in.target_date,
        ),
    )
    conn.commit()
    goal_id = cursor.lastrowid

    cursor.execute("SELECT * FROM goals WHERE id = ?", (goal_id,))
    row = cursor.fetchone()
    conn.close()
    return dict(row)


@router.get("/{goal_id}", response_model=Dict[str, Any])
async def get_goal(
    goal_id: int,
    current_user: Dict[str, Any] = Depends(get_current_user),
) -> Dict[str, Any]:
    user_id = current_user["id"]
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM goals WHERE id = ? AND user_id = ?", (goal_id, user_id))
    row = cursor.fetchone()
    conn.close()

    if not row:
        raise HTTPException(status_code=404, detail=f"Goal with ID {goal_id} not found")

    return dict(row)


@router.patch("/{goal_id}", response_model=Dict[str, Any])
async def update_goal(
    goal_id: int,
    goal_in: GoalUpdateRequest,
    current_user: Dict[str, Any] = Depends(get_current_user),
) -> Dict[str, Any]:
    user_id = current_user["id"]
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM goals WHERE id = ? AND user_id = ?", (goal_id, user_id))
    row = cursor.fetchone()

    if not row:
        conn.close()
        raise HTTPException(status_code=404, detail=f"Goal with ID {goal_id} not found")

    existing_data = dict(row)
    updated_title = goal_in.title if goal_in.title is not None else existing_data["title"]
    updated_desc = goal_in.description if goal_in.description is not None else existing_data["description"]
    updated_cat = goal_in.category if goal_in.category is not None else existing_data["category"]
    updated_status = goal_in.status if goal_in.status is not None else existing_data["status"]
    updated_date = goal_in.target_date if goal_in.target_date is not None else existing_data["target_date"]

    cursor.execute(
        """
        UPDATE goals
        SET title = ?, description = ?, category = ?, status = ?, target_date = ?
        WHERE id = ? AND user_id = ?
        """,
        (updated_title, updated_desc, updated_cat, updated_status, updated_date, goal_id, user_id),
    )
    conn.commit()

    cursor.execute("SELECT * FROM goals WHERE id = ?", (goal_id,))
    updated_row = cursor.fetchone()
    conn.close()
    return dict(updated_row)


@router.delete("/{goal_id}")
async def delete_goal(
    goal_id: int,
    current_user: Dict[str, Any] = Depends(get_current_user),
) -> Dict[str, Any]:
    user_id = current_user["id"]
    conn = get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM goals WHERE id = ? AND user_id = ?", (goal_id, user_id))
    row = cursor.fetchone()

    if not row:
        conn.close()
        raise HTTPException(status_code=404, detail=f"Goal with ID {goal_id} not found")

    # Decouple missions from goal before deletion
    cursor.execute("UPDATE missions SET goal_id = NULL WHERE goal_id = ? AND user_id = ?", (goal_id, user_id))
    cursor.execute("DELETE FROM goals WHERE id = ? AND user_id = ?", (goal_id, user_id))
    conn.commit()
    conn.close()

    return {"message": f"Goal {goal_id} deleted successfully"}

File: netlify/functions/src/controllers/goals.ts
import { toInt } from '../utils/helpers';
import { Router, Response } from 'express';
import { query } from '../db';
import { AuthenticatedRequest, requireAuth } from '../middleware/auth';

const router = Router();

// ============================================================================
// GOALS ENDPOINTS (/api/goals)
// ============================================================================

router.get('/', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const result = await query(
      `SELECT id, title, description, category, status, target_date, blueprint_id, milestone_id, created_at 
       FROM goals 
       WHERE user_id = $1 
       ORDER BY id ASC`,
      [req.user!.id]
    );
    return res.json({ goals: result.rows });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to fetch goals' });
  }
});

router.post('/', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { title, description, category, status, target_date, blueprint_id, milestone_id } = req.body || {};
    if (!title || !title.trim()) {
      return res.status(400).json({ detail: 'Goal title is required.' });
    }

    const result = await query(
      `INSERT INTO goals (user_id, title, description, category, status, target_date, blueprint_id, milestone_id)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       RETURNING id, title, description, category, status, target_date, blueprint_id, milestone_id, created_at`,
      [
        req.user!.id,
        title.trim(),
        description ? description.trim() : '',
        category || 'general',
        status || 'active',
        target_date || null,
        blueprint_id || null,
        milestone_id || null,
      ]
    );

    return res.status(201).json({ goal: result.rows[0] });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to create goal' });
  }
});

router.get('/:goalId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const goalId = toInt(req.params.goalId);
    const result = await query(
      `SELECT id, title, description, category, status, target_date, blueprint_id, milestone_id, created_at 
       FROM goals 
       WHERE id = $1 AND user_id = $2`,
      [goalId, req.user!.id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ detail: 'Goal not found' });
    }

    return res.json({ goal: result.rows[0] });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to fetch goal' });
  }
});

router.patch('/:goalId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const goalId = toInt(req.params.goalId);
    const { title, description, category, status, target_date } = req.body || {};
    const updates: string[] = [];
    const params: any[] = [];
    let pIdx = 1;

    if (title !== undefined) {
      updates.push(`title = $${pIdx++}`);
      params.push(title.trim());
    }
    if (description !== undefined) {
      updates.push(`description = $${pIdx++}`);
      params.push(description.trim());
    }
    if (category !== undefined) {
      updates.push(`category = $${pIdx++}`);
      params.push(category);
    }
    if (status !== undefined) {
      updates.push(`status = $${pIdx++}`);
      params.push(status);
    }
    if (target_date !== undefined) {
      updates.push(`target_date = $${pIdx++}`);
      params.push(target_date);
    }

    if (updates.length === 0) {
      return res.json({ message: 'No updates provided' });
    }

    params.push(goalId, req.user!.id);
    const sql = `UPDATE goals SET ${updates.join(', ')} WHERE id = $${pIdx++} AND user_id = $${pIdx} RETURNING id, title, description, category, status, target_date, blueprint_id, milestone_id, created_at`;
    const result = await query(sql, params);

    if (result.rows.length === 0) {
      return res.status(404).json({ detail: 'Goal not found' });
    }

    return res.json({ goal: result.rows[0] });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to update goal' });
  }
});

router.delete('/:goalId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const goalId = toInt(req.params.goalId);
    await query('DELETE FROM goals WHERE id = $1 AND user_id = $2', [goalId, req.user!.id]);
    return res.json({ message: 'Goal deleted successfully' });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to delete goal' });
  }
});

export default router;
